plugins/net_plugin/include/eosio/net_plugin/net_utils.hpp

```diff
@@ -40,29 +40,71 @@ namespace detail {
       return block_sync_rate_limit;
    }
 
+   /// @return host, port, remainder
+   inline std::tuple<std::string, std::string, std::string> split_host_port_remainder(const std::string& peer_add, bool should_throw) {
+      using std::string;
+      // host:port[:trx|:blk][:<rate>]
+      if (peer_add.empty()) return {};
+
+      auto colon_count = std::count(peer_add.begin(), peer_add.end(), ':');
+      string::size_type end_bracket = 0;
+      if (peer_add[0] == '[') {
+         end_bracket = peer_add.find(']');
+         if (end_bracket == string::npos) {
+            EOS_ASSERT(!should_throw, chain::plugin_config_exception,
+                       "Invalid address specification ${a}, IPv6 no closing square bracket", ("a", peer_add) );
+            return {};
+         }
+      } else if (colon_count >= 7) {
+         EOS_ASSERT(!should_throw, chain::plugin_config_exception,
+                    "Invalid address specification ${a}; IPv6 addresses must be enclosed in square brackets.", ("a", peer_add));
+         return {};
+
+      }
+      string::size_type colon = peer_add.find(':', end_bracket+1);
+      if (colon == string::npos || colon == 0) {
+         return {};
+      }
+      string::size_type colon2 = peer_add.find(':', colon + 1);
+      string host = (end_bracket > 0) ? peer_add.substr( 0, end_bracket+1 ) : peer_add.substr( 0, colon );
+      string port = peer_add.substr( colon + 1, colon2 == string::npos ? string::npos : colon2 - (colon + 1));
+      string remainder = colon2 == string::npos ? "" : peer_add.substr( colon2 + 1 );
+      return {std::move(host), std::move(port), std::move(remainder)};
+   }
+
 } // namespace detail
 
-   /// @return listen address and block sync rate limit (in bytes/sec) of address string
+   /// @return host, port, type. returns empty on invalid peer_add, does not throw
+   inline std::tuple<std::string, std::string, std::string> split_host_port_type(const std::string& peer_add) {
+
+      using std::string;
+      // host:port[:trx|:blk][:<rate>]   // rate is discarded
+      if (peer_add.empty()) return {};
+
+      constexpr bool should_throw = false;
+      auto [host, port, remainder] = detail::split_host_port_remainder(peer_add, should_throw);
+      if (host.empty()) return {};
+
+      string::size_type end = remainder.find_first_of( " :+=.,<>!$%^&(*)|-#@\t" ); // future proof by including most symbols without using regex
+      std::string type = remainder.substr(0, end);
```

Comment:
What is the benefit of doing it this way vs:

```
      std::string type;
      if (remainder.starts_with("blk") || remainder.starts_with("trx"))
         type = remainder.substr(0, 3);
```

Reply:
https://github.com/AntelopeIO/spring/pull/1115/commits/6c2442b14f7724c44625cc8b8f862b1cada1449d